zcash_client_sqlite/src/wallet.rs

```diff
@@ -975,13 +980,183 @@ fn subtree_scan_progress(
+                // shown to users, and gets more accurate the closer to the chain tip we
+                // have scanned.
+                //
+                // TODO: it would be nice to be able to reliably have the size of the
+                // commitment tree at the chain tip without having to have scanned that
+                // block.
+
+                // Get the last scanned height.
+                let last_scanned = conn
+                    .query_row(
+                        "SELECT block_range_end
+                        FROM scan_queue
+                        WHERE priority = :priority
+                        ORDER BY block_range_start DESC
+                        LIMIT 1",
+                        named_params![":priority": priority_code(&ScanPriority::Scanned)],
+                        |row| {
+                            // Scan ranges are end-exclusive.
+                            Ok(BlockHeight::from_u32(row.get(0)?) - 1)
+                        },
+                    )
+                    // `None` if we haven't scanned anything.
+                    .optional()?;
+
+                // Get the tree size at the last scanned height, if known.
+                let last_scanned_tree_size = last_scanned
+                    .map(|last_scanned| {
+                        stmt_end_tree_size_at
+                            .query_row(named_params! {":height": u32::from(last_scanned)}, |row| {
+                                row.get::<_, Option<u64>>(0)
+                            })
+                    })
+                    .transpose()?
+                    .flatten();
+
+                // Get the last completed subtree.
+                let last_completed_subtree = conn
+                    .query_row(
+                        &format!(
+                            "SELECT shard_index, subtree_end_height
+                            FROM {table_prefix}_tree_shards
+                            WHERE subtree_end_height IS NOT NULL
+                            ORDER BY shard_index DESC
+                            LIMIT 1"
+                        ),
+                        [],
+                        |row| Ok((row.get::<_, u64>(0)?, BlockHeight::from_u32(row.get(1)?))),
+                    )
+                    // `None` if we have no subtree roots yet.
+                    .optional()?;
+
+                if let Some((last_completed_subtree_index, last_completed_subtree_height)) =
+                    last_completed_subtree
+                {
+                    // If we know the tree size at the last scanned height, extrapolate.
+                    let tip_tree_size = last_scanned.zip(last_scanned_tree_size).and_then(
+                        |(last_scanned, last_scanned_tree_size)| {
+                            let scanned_notes = last_scanned_tree_size
+                                - (last_completed_subtree_index << shard_height);
```

Comment:
This using `last_completed_subtree_index` seems wrong to me, because we don't have any guarantee at this layer of scanning order; `last_scanned_tree_size` could be far smaller than the last completed subtree. Am I missing something that ties these two together? We need to extrapolate based upon the block position of the last scanned block _within its associated subtree_, but that's not what this is doing except by chance, AFAICT.

Reply:
Indeed I think this is a bug caused by me spending an entire day trying to figure this out and finishing up at 5am.

However, I think the bug is different: if `last_scanned` is _not_ in the most recently-completed subtree, then I think we should extrapolate from the last completed subtree, and _not_ from `last_scanned` (which in that case would be older). We want to extrapolate from the most recently-available data, which is where I went astray here: I missed the condition that `last_scanned > last_completed_subtree_height` here.

Reply:
It's not surprising that this took a long time to figure out; I have a branch where I started attempting to do this sort of interpolation and I abandoned it because it was taking too long. This PR overall is a better approach than I was taking, in that it's an incremental step in the right direction - I was trying to come up with a solution that didn't distinguish between wallet recovery and the steady state and it was a mess.

Reply:
Extrapolating from the previous subtree seems fine; I really wish we just got the tree sizes when we got the chain tip information, that would make it trivial.